### ISSUE TITLE
When incident has lat, long assign place that matches neighborhood

### ISSUE THREAD

**Issue by kidbombay** [Seed-Inc/seed_web#34]

**Comment by jhankin** [comment]
For context: The idea here is that when we do the reverse-geocode based on lat/long, we should run it through the [Google Reverse Geocoding API](https://developers.google.com/maps/documentation/geocoding/intro#ReverseGeocoding), which will give us nested nodes of increasing granularity that correspond to our Place object.  At that point, we should find_or_create a Place representing the neighborhood (or other division larger than an address but smaller than a city) and all of its parents, and link the Incident to that Place.  This way we get all of our incidents nicely organized within a tree of Places, for easy searching via chatbot.

**Comment by kidbombay** [comment]
We should take into account a couple of things
1 - API Down or limit reached. Maybe a rake task that looks at all incidents without places every night and runs find place on them.
2 - Running the assign place as a job. This should not happen at incident creation time. The API requests to external services will slow it down too much.

**Comment by seansu4you87** [comment]
@kidbombay I'm looking into this reverse geocode stuff and saw that you added commit https://github.com/Raheem-ai/seed_web/commit/a84815ba76079d6ceecb0700ef1470929569057c.  The `reverse_geocoded_by` will actually make an external http request, and it looks like this happens after each `save` hook on the `Incident` model.

Also, the text form address isn't enough of what we want (based on what we discussed at the last in person meeting).  We wanted that array of regions in decreasing geographic specificity, in order to create a `Place` model for each region right?  And then we were going to add columns in the `Incident` model for each region area, is that right?

If we want to do the above, then we should also remove the `reverse_geocoded_by`, and then make some async jobs to manually reverse geocode and apply the Place models.

However, if we can think of a shortcut by using the geocoder gem's `reverse_geocoded_by`, I'm happy to stick with that, and take the performance hit of a synchronous external http request.

What do you think?

**Comment by kidbombay** [comment]
It should only make this request if lat/long changes?
If it doesn't that's lame. I just knocked out something quick that calls an existing geocoding from the geocoder gem. Def should optimize to run as job then.

**Comment by kidbombay** [comment]
Also the geocoding gem does have methods to get more than full address. You can break it down from the address components.

**Comment by seansu4you87** [comment]
So here are the components that come back from a reverse geocoding result: ```
Number
Street
Neighborhood
City
Sub State (County)
State
Postal Code
Country
```

Everything above number/street is something we definitely want to index on.  I think there are two main options:

1.  Use the current structure of `Place`, and bubble up parent-child nesting.  Then add columns into each incident to tie each level to the corresponding place
2.  Modify the `Place` data structure to include all of the above levels as column fields, and then keep the tie of one place per incident.  The `Place` row will be able to capture all levels of granularity by appropriate null values.

2 seems cleaner to me, what are your thoughts @kidbombay @jhankin.  If we go with option 2, do I need to worry about current data, or can I just blow everything away?

**Comment by jhankin** [comment]
Is 2 performant?  By my read, we'll have a `Place` for every neighborhood but no umbrella `Place`s for the parents.  It seems to me like it would be expensive if I said I wanted all the incidents in California, and then you had to do a fancy join to get every neighborhood in California, and then every incident that had a place_id in that set.  Pls correct me if I'm not understanding.

**Comment by kidbombay** [comment]
I like 1

**Comment by seansu4you87** [comment]
@jhankin yup u are right, brain fart, thought we were querying for places lol, not incidents

**Comment by jhankin** [comment]
@seansu4you87 word.  One other thought: do we need separate `neighborhood_id` `state_id` etc constants, or can we just have a to-many relationship from `Incident` to `Place`?  (Or a textual array of `place_id`s?  [Is that how a to-many is implemented automagically in Rails?])

**Comment by seansu4you87** [comment]
so for 1:

USA -> California -> San Francisco County -> San Francisco -> Neighborhood

And neighborhood will be the most granular we get.  I'll save the fully formatted address in the Incident.

As for a many-to-many relationship, I like that idea.  Basically incident queries will have/find a place.  And then we can look in the join table for all rows that match the place id, and grab the associated incidents.
